### PR TITLE
docs: add ivy opt-in docs

### PR DESCRIPTION
**Pull request by filipesilva**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?
Adds a Ivy opt-in instructions.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: aio/content/cli/ivy.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Using Ivy with Angular CLI
```
Comment (filipesilva) [inline aio/content/cli/ivy.md]: What is the best place to add this page in?

File: aio/content/cli/ivy.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+{@a updating}
+## Updating an existing project to use Ivy
+
+By configuring a few key files, you can also update your existing project to use Ivy.
+
+- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
+```json
+{
+  "compilerOptions": { ... },
```
Comment (filipesilva) [inline aio/content/cli/ivy.md]: Is using an ellipsis an ok way of representing hidden content for partial options?

File: aio/content/cli/ivy.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+```json
+{
+  "projects": {
+    "my-existing-project": {
+      "architect": {
+        "build": {
+          "options": {
+            ...
+            "aot": true,
+```
```
Comment (filipesilva) [inline aio/content/cli/ivy.md]: Is it better to close all the brackets, or to not use the `json` markdown hint?

File: aio/content/cli/ivy.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+## Updating an existing project to use Ivy
+
+By configuring a few key files, you can also update your existing project to use Ivy.
+
+- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
+```json
+{
+  "compilerOptions": { ... },
+  "angularCompilerOptions": {
+    "enableIvy": true,
```
Comment (cexbrayat) [inline aio/content/cli/ivy.md]: Is true the recommended value here?  The value currently is `ngtsc` with the `experimentalIvy` flag

Comment (gkalpak) [reply aio/content/cli/ivy.md]: Yeah, I don't think `true` works right now. It should be `'ngtsc'`.

Comment (filipesilva) [reply aio/content/cli/ivy.md]: The intend is that it works with `true/false`. I have another task to make those changes.

Comment (gkalpak) [reply aio/content/cli/ivy.md]: 👍 Let's just make sure we don't merge this doc until that other task has been completed 😉

File: aio/content/cli/ivy.md (hunk 1 above)
Comment (IgorMinar) [reply aio/content/cli/ivy.md]: not sure. @jenniferfell can you please propose some place?

File: aio/content/cli/ivy.md (hunk 2 above)
Comment (IgorMinar) [reply aio/content/cli/ivy.md]: yes

File: aio/content/cli/ivy.md (hunk 4 above)
Comment (IgorMinar) [reply aio/content/cli/ivy.md]: true/false is what we want, as @filipesilva mentioned we are landing a change to use simple boolean values for the flag.

File: aio/content/cli/ivy.md (hunk 3 above)
Comment (IgorMinar) [reply aio/content/cli/ivy.md]: close all brackets plz

File: aio/content/cli/ivy.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+
+{@a ngcc}
+## The Angular Compatibility Compiler
+
+Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
+That's where the Angular Compatibility compiler comes into play.
+
+We set it as a postinstall script in your `package.json` so that it always runs when you install
+new modules.
```
Comment (IgorMinar) [inline aio/content/cli/ivy.md]: Please add: Using postinstall hook to run ngcc is just a temporary integration. We expect ngcc to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. Once that's implemented ngcc will and not be visible to developers.

**Review by IgorMinar — APPROVED** [review]
great start!

Let's wait for the blog post to land and experimentalIvy boolean flag to land and then merge this PR

**Comment by filipesilva** [comment]
Blocked on https://github.com/angular/angular/pull/28616.

File: aio/content/cli/ivy.md (hunk 5 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Done.

File: aio/content/cli/ivy.md (hunk 3 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Done

File: aio/content/cli/ivy.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
+That's where the Angular Compatibility (`ngcc`) compiler comes into play.
+
+We set it as a postinstall script in your `package.json` so that it always runs when you install
+new modules.
+
+Using a postinstall hook to run `ngcc` is just a temporary integration. 
+We expect `ngcc` to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. 
+Once that's implemented `ngcc` will and not be visible to developers.
```
Comment (alan-agius4) [inline aio/content/cli/ivy.md]: @filipesilva, this is really good!

Should we mentioned the limitation that if one has multiple projects within the workspace, all of them will need to be build using Ivy? As users will have cryptic error messages on projects that rely on `ngc` but `ngcc` was run.

File: aio/content/cli/ivy.md (hunk 4 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: https://github.com/angular/angular/pull/28616 is now merged.

File: aio/content/cli/ivy.md (hunk 6 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Added a note about that:

> Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.

**Comment by filipesilva** [comment]
@jenniferfell can you review and advise on where the best place to link this page in is please?

File: aio/content/cli/ivy.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+## Updating an existing project to use Ivy
+
+By configuring a few key files, you can also update your existing project to use Ivy.
+
+- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
+```json
+{
+  "compilerOptions": { ... },
+  "angularCompilerOptions": {
+    "enableIvy": true,
```
Comment (gkalpak) [inline aio/content/cli/ivy.md]: Isn't `"allowEmptyCodegenFiles": true` required any more?

Comment (filipesilva) [reply aio/content/cli/ivy.md]: Yes, it is still required right now. But https://github.com/angular/angular/pull/28685 (or a similar approach) will make it not needed.

Comment (filipesilva) [reply aio/content/cli/ivy.md]: Added it here for now, to remove later when the other PR is in.

File: aio/content/cli/ivy.md (hunk 1 above)
Comment (brandonroberts) [reply aio/content/cli/ivy.md]: This doc should be placed in the `aio/content/guide` directory. I think it would be good also to rename the doc to something like `ivy-preview` if this is going to be temporary.

File: aio/content/cli/ivy.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+# Using Ivy with Angular CLI
+
+Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).
+
+
+## Starting a new project using Ivy
+
+The easiest way to start a new project with Ivy is to use the `--experimental-ivy` flag on `ng new`:
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
To start a new project with Ivy enabled, use the `--experimental-ivy` flag with the [`ng new`](cli/new) command:
```

File: aio/content/cli/ivy.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+```
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
```sh
```

File: aio/content/cli/ivy.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ng new shiny-ivy-app --experimental-ivy
+```
+
+Everything will be configured for you.
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]: I think this should list the steps that will be done similar to the list of items that you have to do to undo the changes.

File: aio/content/cli/ivy.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+{@a updating}
+## Updating an existing project to use Ivy
+
+By configuring a few key files, you can also update your existing project to use Ivy.
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
By configuring a few key files, you can also update your existing project(s) to use Ivy.
```

File: aio/content/cli/ivy.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
- Add the `enableIvy` option in the `angularCompilerOptions` in your project's `src/tsconfig.app.json`:
```

File: aio/content/cli/ivy.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+```json
+{
+  "compilerOptions": { ... },
+  "angularCompilerOptions": {
+    "enableIvy": true,
+    "allowEmptyCodegenFiles": true,
+  }
+}
+```
+- set the default build options for your project to always use AOT compilation if it isn't already:
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
- Set the default build options for your project to always use AOT compilation if it isn't already:
```

File: aio/content/cli/ivy.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            ...
+            "aot": true,
+          }
+        }
+      }
+    }
+  }
+}
+```
+- add a `postinstall` script in `package.json` to run the [Angular Compatibility Compiler](#ngcc):
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
- Add a `postinstall` script to the workspace `package.json` file to run the [Angular Compatibility Compiler](#ngcc):
```

File: aio/content/cli/ivy.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+```json
+{
+  "scripts": {
+    ...
+    "postinstall": "ivy-ngcc"
+  }
+}
+```
+- reinstall your node modules to run the newly added script.
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
- Reinstall your `package.json` dependencies to run the newly added script.
```

File: aio/content/cli/ivy.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+## Switching back to the current compiler
+
+To stop using the Ivy compiler you need to undo the steps taken when [updating to use Ivy](#updating).
+- set `enableIvy` to false in `src/tsconfig`, or remove it completely.
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]: Capitalize each bullet

File: aio/content/cli/ivy.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+
+{@a ngcc}
+## The Angular Compatibility Compiler
+
+Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
+That's where the Angular Compatibility (`ngcc`) compiler comes into play.
+
+We set it as a postinstall script in your `package.json` so that it always runs when you install
+new modules.
```
Comment (brandonroberts) [inline aio/content/cli/ivy.md]:
```suggestion
install your `package.json` dependencies.
```

**Review by brandonroberts — DISMISSED** [review]

File: aio/content/cli/ivy.md (hunk 1 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Renamed and added under the `Techniques` section of navigation as `Ivy Preview`. @StephenFluin  @jenniferfell can you confirm this location is the appropriate one?

File: aio/content/cli/ivy.md (hunk 16 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Done

File: aio/content/cli/ivy.md (hunk 10 above)
Comment (filipesilva) [reply aio/content/cli/ivy.md]: Done.

File: aio/content/guide/ivy-preview.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
+# Previewing Ivy with Angular CLI
+
+Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]:
```suggestion
Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](https://blog.angular.io/a-plan-for-version-8-0-and-ivy-b3318dfc19f7).
```

File: aio/content/guide/ivy-preview.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+## Starting a new project using Ivy
+
+To start a new project with Ivy enabled, use the `--experimental-ivy` flag with the [`ng new`](cli/new) command:
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]: did we notice that the flag name is different between CLI and tsconfig? "experimental-ivy" vs "enableIvy" - could we consolidate these?

Comment (IgorMinar) [reply aio/content/guide/ivy-preview.md]: @alexeagle @kara @alxhub @filipesilva ⬆️

File: aio/content/guide/ivy-preview.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+By configuring a few key files, you can also update your existing project(s) to use Ivy.
+
+- Add the `enableIvy` option in the `angularCompilerOptions` in your project's `src/tsconfig.app.json`:
+```json
+{
+  "compilerOptions": { ... },
+  "angularCompilerOptions": {
+    "enableIvy": true,
+    "allowEmptyCodegenFiles": true,
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]:
```suggestion
    # you might also need to enable shim file generation
    # this requirement should go away before version 8 ships
    "allowEmptyCodegenFiles": true,
```

Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]: @kara @alxhub  we should create a flag with a human understandable name that is related to shim generation instead of piggybacking on this one, e.g. "generateNgFactoryShimFiles" or similar.

File: aio/content/guide/ivy-preview.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+    }
+  }
+}
+```
+- Add a `postinstall` script to the workspace `package.json` file to run the [Angular Compatibility Compiler](#ngcc):
+```json
+{
+  "scripts": {
+    ...
+    "postinstall": "ivy-ngcc"
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]: @petebacondarwin @gkalpak 
@kara @alxhub  `ivy-ngcc` is an odd looking binary filename. why not just call it `ngcc`?

File: aio/content/guide/ivy-preview.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+  }
+}
+```
+- Reinstall your `package.json` dependencies to run the newly added script.
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]:
````suggestion
- Reinstall your `package.json` dependencies to run the newly added script:

```
npm install
```
````

File: aio/content/guide/ivy-preview.md [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+That's where the Angular Compatibility (`ngcc`) compiler comes into play.
+
+We set it as a postinstall script in your `package.json` so that it always runs when you install
+your `package.json` dependencies.
+
+Using a postinstall hook to run `ngcc` is just a temporary integration. 
+We expect `ngcc` to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. 
+Once that's implemented `ngcc` will and not be visible to developers.
+
+Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]:
```suggestion
Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.

Note: we are aware of build-time performance issues with ngcc. Please bear with us while we improve the tooling and build integration in the Angular CLI.
```

**Review by IgorMinar — APPROVED** [review]
left a few suggestions - if you accept these then LGTM

**Comment by IgorMinar** [comment]
I've dropped @jenniferfell from requested reviewers for now as she's OOO.

@jenniferfell could you please do a post-submit review of this PR? let's get it out as is and iterate on it while it's already on next.angular.io

File: aio/content/guide/ivy-preview.md [hunk 24]
```diff
@@ -0,0 +1,84 @@
+# Previewing Ivy with Angular CLI
```
Comment (IgorMinar) [inline aio/content/guide/ivy-preview.md]:
```suggestion
# Opt-ing into Angular Ivy with Angular CLI
```

File: aio/content/navigation.json [hunk 25]
```diff
@@ -515,22 +502,24 @@
               "tooltip": "Extend Angular by creating, publishing, and using your own libraries."
             }
           ]
+        },
+        {
+          "url": "guide/ivy-preview",
+          "title": "Ivy Preview",
+          "tooltip": "Previewing Ivy with Angular CLI."
```
Comment (IgorMinar) [inline aio/content/navigation.json]:
```suggestion
          "tooltip": "Opt-ing into Angular Ivy with Angular CLI."
```

File: aio/content/navigation.json [hunk 26]
```diff
@@ -515,22 +502,24 @@
               "tooltip": "Extend Angular by creating, publishing, and using your own libraries."
             }
           ]
+        },
+        {
+          "url": "guide/ivy-preview",
+          "title": "Ivy Preview",
```
Comment (IgorMinar) [inline aio/content/navigation.json]:
```suggestion
          "title": "Angular Ivy",
```

File: aio/content/navigation.json [hunk 27]
```diff
@@ -515,22 +502,24 @@
               "tooltip": "Extend Angular by creating, publishing, and using your own libraries."
             }
           ]
+        },
+        {
+          "url": "guide/ivy-preview",
```
Comment (IgorMinar) [inline aio/content/navigation.json]:
```suggestion
          "url": "guide/ivy",
```

Comment (IgorMinar) [inline aio/content/navigation.json]: @filipesilva can you please rename this file to `ivy.md`? that will change the url to the one suggested above.

**Review by IgorMinar — APPROVED** [review]
a few more suggestions to incorporate...

File: aio/content/navigation.json (hunk 27 above)
Comment (filipesilva) [reply aio/content/navigation.json]: Done

File: aio/content/guide/ivy-preview.md (hunk 19 above)
Comment (filipesilva) [reply aio/content/guide/ivy-preview.md]: I think the easiest way to consolidate is to change the CLI `--experimental-ivy` to `--enable-ivy`. 

We could do that, but then we lose the implicit opt-in that goes with a user having to type out the `experimental` word. But that's ok if we commit to keep the `--enable-ivy` flag around and it not being experimental at all.

File: aio/content/guide/ivy-preview.md (hunk 20 above)
Comment (filipesilva) [reply aio/content/guide/ivy-preview.md]: Although `tsconfig.json` itself accepts comments, the `src/tsconfig.app.json` file is not automatically identified by VS Code as being a TypeScript configuration file. This means comments appear as an error.

Instead of adding this comment, I added text to the instruction above the code.

**Comment by filipesilva** [comment]
To summarize, this PR is ready to merge but there are still open action items that should be addressed and whose changes are to be fed back into this document:
- consolidate flags (`--experimental-ivy` vs `--enable-ivy`): https://github.com/angular/angular/pull/28569#discussion_r259558085 
  - AI for @filipesilva to do on CLI repo
- human readable name to replace `allowEmptyCodegenFiles` : https://github.com/angular/angular/pull/28569#discussion_r259558375
  - AI for @kara @alxhub to comment on, TBD implementer
- `ivy-ngcc` binary name: https://github.com/angular/angular/pull/28569#discussion_r259558440
  - AI for @petebacondarwin @gkalpak @kara @alxhub to comment on, TBD implementer

File: aio/content/guide/ivy.md [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+
+Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
+That's where the Angular Compatibility (`ngcc`) compiler comes into play.
+
+We set it as a postinstall script in your `package.json` so that it always runs when you install
+your `package.json` dependencies.
+
+Using a postinstall hook to run `ngcc` is just a temporary integration. 
+We expect `ngcc` to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. 
+Once that's implemented `ngcc` will and not be visible to developers.
```
Comment (benb7760) [inline aio/content/guide/ivy.md]: remove 'and' from this sentence?

Comment (filipesilva) [reply aio/content/guide/ivy.md]: Hey thanks for taking a look! Yes that shouldn't be there, removed now.

File: aio/content/navigation.json [hunk 29]
```diff
@@ -515,22 +502,24 @@
               "tooltip": "Extend Angular by creating, publishing, and using your own libraries."
             }
           ]
+        },
+        {
+          "url": "guide/ivy",
+          "title": "Angular Ivy",
+          "tooltip": "Opt-ing into Angular Ivy with Angular CLI."
```
Comment (petebacondarwin) [inline aio/content/navigation.json]: Opting is a word - no hyphen needed

Comment (filipesilva) [reply aio/content/navigation.json]: I changed mentions of `opt-ing` to `opting` but left mentions of `opt-in` as is.

File: aio/content/guide/ivy-preview.md (hunk 21 above)
Comment (kara) [reply aio/content/guide/ivy-preview.md]: +1 on `ngcc`

Comment (gkalpak) [reply aio/content/guide/ivy-preview.md]: > why not just call it `ngcc`?

Maybe because the next time we need a compatibility compiler, we would have to name it `ngcc2` (a la `Renderer2`), which is not cool 😛
